### PR TITLE
Add `Guild._fetch_role()` private method to support proper handling of role mentionable in slash command options

### LINKED ISSUE #906: slash command option input type "mentionable" issue
### Summary

When you mention role in slash command, exception occures

### Reproduction Steps

run code below and try to mention a role in slash command

### Minimal Reproducible Code

```python
bot = discord.Bot()
@bot.slash_command(guild_ids=["id"])
async def command(
    ctx,
    var: discord.Option((discord.Role, discord.User))
):
    print(var)

bot.run("token")
```

### Expected Results

- if user mentioned in slash command, prints user object. ✅

- if role mentioned in slash command, prints role object. ❌

### Actual Results

- The following traceback is printed:

```
Ignoring exception in command <discord.commands.SlashCommand name=command>:
Traceback (most recent call last):
  File "C:\Users\{username}\AppData\Local\Programs\Python\Python310\lib\site-packages\discord\commands\core.py", line 110, in wrapped
  File "C:\Users\{username}\AppData\Local\Programs\Python\Python310\lib\site-packages\discord\http.py", line 334, in request
    raise NotFound(response, data)
discord.errors.NotFound: 404 Not Found (error code: 10013): Unknown User

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\{username}\AppData\Local\Programs\Python\Python310\lib\site-packages\discord\bot.py", line 520, in process_application_commands
    await ctx.command.invoke(ctx)
  File "C:\Users\{username}\AppData\Local\Programs\Python\Python310\lib\site-packages\discord\commands\core.py", line 306, in invoke
    await injected(ctx)
  File "C:\Users\{username}\AppData\Local\Programs\Python\Python310\lib\site-packages\discord\commands\core.py", line 116, in wrapped
    raise ApplicationCommandInvokeError(exc) from exc
discord.commands.errors.ApplicationCommandInvokeError: Application Command raised an exception: NotFound: 404 Not Found (error code: 10013): Unknown User
```

### Intents

all

### System Information

- Python v3.10.2-final
- py-cord v2.0.0-beta
    - py-cord pkg_resources: v2.0.0b4897+g4dadb69d
- aiohttp v3.8.1
- system info: Windows 10 10.0.22000

### Checklist

- [X] I have searched the open issues for duplicates.
- [X] I have shown the entire traceback, if possible.
- [X] I have removed my token from display, if visible.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by krittick**
## Summary

This adds a new private method `Guild._fetch_role()` which allows `utils.get_or_fetch` to work with roles properly. Previously it did not work because there was only a `fetch_roles` method, not `fetch_role`. As part of this change, `utils.get_or_fetch` was also updated to use private `_fetch` methods if an `AttributeError` is raised for any object that does not have a `fetch_*` method. This has an added benefit of allowing us to more easily add private _fetch methods on other objects if we need to in the future.

This also fixes #906 by having the "mentionable" option type check for first a user, then a role if a user is not found. This required the above described addition of `Guild._fetch_fole()` to work properly.

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why
- [x] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...)

### CODE REVIEW

File: discord/guild.py [hunk 1]
```diff
@@ -2474,6 +2474,38 @@ async def fetch_roles(self) -> List[Role]:
...
+    async def fetch_role(self, role_id: int) -> Role:
+        """|coro|
+
+        Retrieves a :class:`Role` that the guild has.
+
+        .. note::
+
+            This method is an API call. For general usage, consider iterating over :attr:`roles` instead.
+
+        .. versionadded:: 1.3
```
Comment (Dorukyum) [inline discord/guild.py:2486]: I think you forgot to change this :)
```suggestion
        .. versionadded:: 2.0
```

**Review by Dorukyum — CHANGES_REQUESTED** [review]

**Review by Dorukyum — DISMISSED** [review]

**Review by zeffo — COMMENTED** [review]
Perhaps the note should be changed to ```This method is an API call. For general usage, consider using :attr:`get_role` instead.```

**Comment by BobDotCom** [comment]
I feel like making this a private method "_fetch_role" would be more correct (And add handling for that inside get_or_fetch). This currently doesn't reflect the api well and shouldn't really be a method we promote, as it's rather "fetch_roles_and_filter_for_one_role"

**Review by BobDotCom — CHANGES_REQUESTED** [review]
See my last comment

**Comment by krittick** [comment]
> I feel like making this a private method "_fetch_role" would be more correct (And add handling for that inside get_or_fetch). This currently doesn't reflect the api well and shouldn't really be a method we promote, as it's rather "fetch_roles_and_filter_for_one_role"

Good point. I'll update with those changes and test it out. Setting to draft for now.

**Comment by krittick** [comment]
This is ready for review now.

**Review by Icebluewolf — APPROVED** [review]
I downloaded this branch and It fixed the problem for me!

**Review by Dorukyum — APPROVED** [review]

**Review by RGBCube — APPROVED** [review]

**Review by BobDotCom — APPROVED** [review]

**Pull request merged**